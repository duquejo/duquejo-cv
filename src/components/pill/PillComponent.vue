<template>
  <span class="border-yellow-500 border-2 font-semibold rounded hover:translate-y-0.5 transition-transform" :class="elementClasses">{{ props.text }}</span>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import type { Sizes } from '@/interfaces';

interface Props {
  text: string;
  size?: Sizes;
}

const props = withDefaults(defineProps<Props>(), {
  size: 'medium',
});

const elementClasses = computed(() => ({
  'text-xs px-2.5 py-1.5': props.size === 'medium',
  'text-[0.6rem] px-2 py-0.5': props.size === 'small',
}));
</script>