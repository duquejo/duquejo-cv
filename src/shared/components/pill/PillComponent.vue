<template>
  <span class="border-yellow-500 border-2 text-xs font-semibold px-2.5 py-1.5 rounded hover:translate-y-0.5 transition-transform">{{ props.text }}</span>
</template>

<script setup lang="ts">
interface Props {
  text: string;
}
const props = defineProps<Props>();
</script>