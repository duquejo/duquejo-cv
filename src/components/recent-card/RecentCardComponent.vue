
<template>
  <div class="bg-white rounded pl-5 pr-4 py-4 last:mb-8 first:mt-1 mx-2 my-4 shadow flex-col">
    <time v-if="parseDate" class="text-xs font-extrabold text-right">{{ parseDate }}</time>
    <div class="flex items-center text-xs mb-2 justify-between gap-x-5">
      <span class="flex items-center gap-x-2">
        <img :src="event.actor.avatar_url" :alt="`@${event.actor.display_login}`" class="max-h-7 rounded-full" />
        <a class="font-semibold underline underline-offset-2" :href="`https://github.com/${event.repo.name}`" target="_blank">{{ event.repo.name }}</a>
      </span>
      <PillComponent 
        v-if="event.payload.ref && event.type === EventType.CreateEvent" 
        :text="event.payload.ref" 
        :title="event.payload.ref" 
        size="small"
      />
      <PillComponent 
        class="truncate"
        v-else-if="event.payload.pull_request && event.type === EventType.PullRequestEvent" 
        :text="event.payload.pull_request.head.ref" 
        :title="event.payload.pull_request.head.ref"
        size="small"
      />
    </div>
    <p v-if="! [EventType.PullRequestEvent, EventType.WatchEvent].includes(event.type)" class="bg-gray-100 rounded-lg p-3 my- text-xs leading-relaxed text-left">
      <strong>{{ event.payload.description }}</strong>
      <ul v-if="event.payload.commits" class="block">
        <li v-for="commit in event.payload.commits" :key="commit.sha">
          {{ markdownText(commit.message) }}
          <a
            v-if="event.type === 'PushEvent'"
            class="text-[0.7rem] font-semibold text-gray-700"
            target="_blank"
            :href="`${commit.url.replace('https://api.github.com/repos', 'https://github.com')}`"
            :title="t('general.menu.events_repo_commit_link')">
            <v-icon name="bi-github" scale="0.8"/>
          </a>
        </li>
      </ul>
    </p>
    <div data-test="rc-type" class="flex items-center px-2 py-2 rounded-3xl gap-x-3 shadow-md relative hover:translate-y-0.5 transition-transform ring-1 ring-gray-200 text-xs">
      <span class="p-1 rounded-full leading-none max-h-7" :class="icon.color">
        <component is="v-icon" :name="icon.icon" scale="1" />
      </span>
      {{ event.type }}
    </div>
  </div>
</template>

<script setup lang="ts">
import { toRefs, computed } from 'vue';
import { useI18n } from 'vue-i18n';
import * as emoji from 'node-emoji';

import type { Event } from '@/interfaces/github-events';
import { EventType } from '@/interfaces/github-events';
import PillComponent from '@/components/pill/PillComponent.vue';


interface Props {
  event: Event
}

const props = defineProps<Props>();
const { event } = toRefs(props);

const { t } = useI18n();

const markdownText = (text: string): string => emoji.emojify(text);

const icon = computed(() => {
  switch (event.value.type) {
    case EventType.PullRequestEvent:
      return {
        icon: 'io-git-pull-request-sharp',
        color: 'bg-yellow-400',
      };
    case EventType.CreateEvent:
      return {
        icon: 'io-git-branch',
        color: 'bg-teal-400',
      };
    case EventType.PushEvent:
      return {
        icon: 'oi-repo-push',
        color: 'bg-red-400',
      };
    case EventType.WatchEvent: 
      return {
        icon: 'ri-eye-2-line',
        color: 'bg-cyan-300',
      };
    default:
      return {
        icon: 'oi-repo-push'
      };
  };
});

const parseDate = computed(() => {
  const date = new Date(event.value.created_at);

  if( ! ( date instanceof Date && ! isNaN(date.getTime()) ) ) {
    return false
  }

  return date.toLocaleString();
});
</script>