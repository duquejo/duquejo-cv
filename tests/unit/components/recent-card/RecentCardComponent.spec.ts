import { shallowMount } from '@vue/test-utils';
import { describe, expect, it, vi, afterEach, beforeEach, MockInstance } from 'vitest';

import RecentCardComponent from '../../../../src/components/recent-card/RecentCardComponent.vue';
import PillComponent from '../../../../src/components/pill/PillComponent.vue';
import createVueI18n from '../../../mocks/mock-i18n';
import events from '../../../mocks/events.json';

describe('+ RecentCardComponent.vue tests', () => {

  let mockDate: MockInstance<[locales?: Intl.LocalesArgument, options?: Intl.DateTimeFormatOptions], string>;

  const mockedDate = '12/22/2023, 1:00:00 PM';
  const i18n = createVueI18n('en');

  const pushEvent = events[0];
  const createEvent = events[1];
  const pullRequestEvent = events[2];
  const watchEvent = events[3];

  const createShallowMount = (eventType = pushEvent) => shallowMount(RecentCardComponent, {
    props: {
      event: eventType,
    },
    global: {
      plugins: [i18n],
      stubs: {
        PillComponent: PillComponent,
      }
    }
  });
  

  beforeEach(() => {
    mockDate = vi.spyOn(Date.prototype, 'toLocaleString').mockReturnValue(mockedDate);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('Should match the snapshot', () => {

    const wrapper = createShallowMount();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render conditionally its sections based on event type - PushEvent', () => {
    const wrapper = createShallowMount(pushEvent);

    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeSpan = typeContainer.find('span');
    const typeIcon = typeContainer.find('v-icon');

    const a = wrapper.findAll('p a');

    expect(a.length).toBeGreaterThan(0);
    expect(wrapper.findComponent(PillComponent).exists()).toBeFalsy();
    
    expect(typeContainer.text()).toBe(pushEvent.type);
    expect(typeIcon.attributes()).toEqual({ name: 'oi-repo-push', scale: '1' });
    expect(typeSpan.classes()).toContain('bg-red-400');
  });

  it('should render conditionally its sections based on event type - WatchEvent', () => {
    const wrapper = createShallowMount(watchEvent);

    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeSpan = typeContainer.find('span');
    const typeIcon = typeContainer.find('v-icon');

    const a = wrapper.find('p a');

    expect(a.exists()).toBeFalsy();
    expect(wrapper.findComponent(PillComponent).exists()).toBeFalsy();
    
    expect(typeContainer.text()).toBe(watchEvent.type);
    expect(typeIcon.attributes()).toEqual({ name: 'ri-eye-2-line', scale: '1' });
    expect(typeSpan.classes()).toContain('bg-cyan-300');
  });

  it('should render conditionally its sections based on event type - CreateEvent', () => {
    const wrapper = createShallowMount(createEvent);

    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeSpan = typeContainer.find('span');
    const typeIcon = typeContainer.find('v-icon');

    const a = wrapper.find('p a');

    expect(a.exists()).toBeFalsy();
    expect(wrapper.findComponent(PillComponent).exists()).toBeTruthy();

    expect(typeContainer.text()).toContain(createEvent.type);
    expect(typeIcon.attributes()).toEqual({ name: 'io-git-branch', scale: '1' });
    expect(typeSpan.classes()).toContain('bg-teal-400');
  });

  it('should render conditionally its sections based on event type - PullRequestEvent', () => {
    const wrapper = createShallowMount(pullRequestEvent);

    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeSpan = typeContainer.find('span');
    const typeIcon = typeContainer.find('v-icon');

    const a = wrapper.find('p a');

    expect(a.exists()).toBeFalsy();
    expect(wrapper.findComponent(PillComponent).exists()).toBeTruthy();

    expect(typeContainer.text()).toBe(pullRequestEvent.type);
    expect(typeIcon.attributes()).toEqual({ name: 'io-git-pull-request-sharp', scale: '1' });
    expect(typeSpan.classes()).toContain('bg-yellow-400');
  });

  
  it('should render conditionally its sections based on event type - Default fallback', () => {
    const wrapper = createShallowMount({
      ...watchEvent,
      type: 'FooBar'
    });

    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeIcon = typeContainer.find('v-icon');

    expect(typeIcon.attributes()).toEqual({ name: 'oi-repo-push', scale: '1' });
  });


  it('should render conditionally the date tag', () => {
    const wrapper = createShallowMount({
      ...pushEvent,
      created_at: 'foo-bar', // Invalid date
    });

    const date = wrapper.find('time');
    const typeContainer = wrapper.find('[data-test="rc-type"]');
    const typeIcon = typeContainer.find('v-icon');

    expect(typeIcon.attributes()).toEqual({ name: 'oi-repo-push', scale: '1' });
    expect(date.exists()).toBeFalsy();
  });
});