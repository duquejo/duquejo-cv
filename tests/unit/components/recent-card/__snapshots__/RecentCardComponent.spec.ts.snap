// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ RecentCardComponent.vue tests > Should match the snapshot 1`] = `
"<div class="bg-white rounded pl-5 pr-4 py-4 last:mb-8 first:mt-1 mx-2 my-4 shadow flex-col">
  <date class="text-xs font-extrabold text-right">12/22/2023, 1:05:37 PM</date>
  <div class="flex items-center text-xs mb-2 justify-between gap-x-5"><span class="flex items-center gap-x-2"><img src="https://avatars.githubusercontent.com/u/47703424?" alt="@duquejo01" class="max-h-7 rounded-full"><a class="font-semibold underline underline-offset-2" href="https://github.com/duquejo01/vue-cv" target="_blank">duquejo01/vue-cv</a></span>
    <!--v-if-->
  </div>
  <p class="bg-gray-100 rounded-lg p-3 my- text-xs leading-relaxed text-left"><strong></strong>
  <ul class="block">
    <li>âœ¨ Translation manager helper class <a class="text-[0.7rem] font-semibold text-gray-700" target="_blank" href="https://github.com/duquejo01/vue-cv/commits/9e288ef7b8ade3c0f1be25902f08e0754ea479b4" title="general.menu.events_repo_commit_link">
        <v-icon name="bi-github" scale="0.8"></v-icon>
      </a></li>
  </ul>
  </p>
  <div data-test="rc-type" class="flex items-center px-2 py-2 rounded-3xl gap-x-3 shadow-md relative hover:translate-y-0.5 transition-transform ring-1 ring-gray-200 text-xs"><span class="p-1 rounded-full leading-none max-h-7 bg-red-400"><v-icon name="oi-repo-push" scale="1"></v-icon></span> PushEvent</div>
</div>"
`;
